// @accessors-modules/gdp/gdp.js

// Copyright (c) 2016 The Regents of the University of California.
// All rights reserved.

// Permission is hereby granted, without written agreement and without
// license or royalty fees, to use, copy, modify, and distribute this
// software and its documentation for any purpose, provided that the above
// copyright notice and the following two paragraphs appear in all copies
// of this software.

// IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
// FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
// ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
// THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
// SUCH DAMAGE.

// THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE
// PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND THE UNIVERSITY OF
// CALIFORNIA HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
// ENHANCEMENTS, OR MODIFICATIONS.

/**
 * Module supporting the Global Data Plane.
 *
 * @module @accessors-modules/gdp
 * @author Christopher Brooks
 * @version $$Id$$
 */

// Stop extra messages from jslint.  Note that there should be no
// space between the / and the * and global.
/*globals exports, Java, require, util */
/*jshint globalstrict: true */
'use strict';

var gdpjs = require('@terraswarm/gdp');
var EventEmitter = require('events').EventEmitter;
var util = require('util');

var gclH;
var os = require('os');

var debug = false;

/** Instantiate a GDPHelper.  Create the log if necessary.
 *  @param name The name of the GDP log.  Dot-separated reverse notation preferred: edu.berkeley.terraswarm.yourname.log00
 *  @param iomode Opening mode (0: for internal use only, 1: read-only, 2: read-append, 3: append-only)
 * @param logdName  Name of the log server where this should be placed if it does not yet exist.
 */
exports.GDP = function (name, iomode, logdname) {
    this.gclh = gdpjs.gdp_gcl_open(name, iomode, logdname);
    this.name = name;
}

util.inherits(exports.GDP, EventEmitter);

/** Append data to the already open log. 
 *  @param {string} data The data to be appended
 */
exports.GDP.prototype.append = function (data) {
    // FIXME: gdpdAddress should be a parameter.
    var gdpdAddress = null;
    var gclName = this.name;
    var logdxname = os.hostname();
    var gclAppend = true;
    var recordSource = 0;
    var recordArray = new Array(data);
    var consoleOut = false;
    var recordArrayOut = new Array(1);
    var returnValue = gdpjs.write_gcl_records(gdpdAddress, gclName, logdxname, gclAppend,
                                              recordSource, recordArray, consoleOut,
                                              recordArrayOut);
    if (!returnValue.error_isok) {
        throw new Error("gdp.js: append() failed: " + returnValue.error_msg + ", gcl_name: " + returnValue.gclName);
    }
};

/** Close the gcl. */
exports.GDP.prototype.close = function () {
    gdpjs.gdp_gcl_close_js(this.gclh);
}

/** Read a record.
 *  @param {int} recno The record number.  GDP records start with 1.
 *  @return {string} The data.
 */
exports.GDP.prototype.read = function (recno) {
    if (debug) {
        console.log("node_modules/gdp/gdp.js:read(" + recno + ")");
    }
    var gdpd_addr = gdpd_addr;
    var gcl_name = this.name
    var gcl_firstrec = recno;
    var gcl_numrecs = 1;
    var gcl_subscribe = false;
    var gcl_multiread = false;
    var conout = true; // recdest = -1 so output to console.log()
    var gdp_event_cbfunc = null; // no callback needed when just logging
    var wait_for_events = true; // wait indefinitely in read_gcl_records() for
    var gcl_get_next_event = false;
    var data = gdpjs.read_gcl_records(gdpd_addr, gcl_name,
                                      gcl_firstrec, gcl_numrecs,
                                      gcl_subscribe, gcl_multiread, /*recdest,*/
                                      conout, gdp_event_cbfunc,
                                      wait_for_events,
                                      gcl_get_next_event
                                     );
    if (data.records.length > 0) {
        return data.records[0].value;
    } else {
        return null;
    }
};

/** Set the debug level.
 *  @param {string} debugLevel The debug level.  "*=10" will set the
 *  level to 10 for all modules.  See gdp/README-developers.md for
 *  details.  The value is typically
 *  <code><i>pattern</i>=<i>level</i></code>, for example
 *  <code>gdplogd.physlog=39</code>.  To see the patterns, use the
 *  "what" command or <code>strings $PTII/lib/libgdp* | grep
 *  '@(#)'</code>.  Use <code>*=40</code> to set the debug level to 40
 *  for all components. The value of level is not usually over 127.
 *  Values over 100 may modify the behavior.
 */
exports.GDP.prototype.setDebugLevel = function (debugLevel) {
    if (debug) {
        console.log("gdp.js: setDebugLevel(" + debugLevel + ")");
    }
    gdpjs.ep_dbg_set(debugLevel)
}

/** Keep track of subscription callbacks
 * http://stackoverflow.com/questions/36819679/node-fficall-c-lib-with-callback-function
 * says that we need to keep a reference to what is returned from ffi.Callback so as to avoid garbage collection.
 */
var  _subscriptionCallBacks = [];

/** Subscribe.
 *  @param self The CapeCode subscribe() function passes "this" as its first argument,
 *  so we need to do the same.
 *  @param {int} startrec The starting record. The first record is record 1.
 *  @param {int} numrecs The number of records
 *  @param {int} timeout The timeout in milliseconds.
 */
exports.GDP.prototype.subscribe = function (self, startrec, numrecs, timeout) {
    //     console.log("node_modules/gdp/gdp.js:subscribe(" + startrec + ", " + numrecs + ", " + timeout + ")");
    var gdpd_addr = gdpd_addr;
    var gcl_name = this.name
    var gcl_firstrec = startrec;
    var gcl_numrecs = numrecs
    var gcl_subscribe = true;
    var gcl_multiread = false;
    var conout = true; // recdest = -1 so output to console.log()
    var wait_for_events = false; // don't wait indefinitely in read_gcl_records() for

    var gcl_get_next_event = false;

    // See https://www.icyphy.org/accessors/wiki/Main/GDPJSCallbacks
    // and http://stackoverflow.com/questions/36819679/node-fficall-c-lib-with-callback-function

    // Define an event callback.

    // gdp.h defines:
    // typedef void (*gdp_event_cbfunc_t)(        // the callback function
    //      gdp_event_t *ev);                           // the event triggering the call

    // FIXME: Copied from gdpjs.sj
    var ffi = require('ffi');
    var ref = require('ref');
    var gdp_event_t = ref.types.void; // opaque for us up here in JS
    var gdp_event_tPtr = ref.refType(gdp_event_t);

    // End of copy

    var gdp_event_cbfunc = null // FIXME: not sure about this: no callback needed when just logging

    // FIXME: http://stackoverflow.com/questions/36819679/node-fficall-c-lib-with-callback-function
    // says that we need to keep a reference to what is returned from ffi.Callback so as to avoid garbage collection.

    var self = this;
    var gdp_event_cbfunc = ffi.Callback('void', [gdp_event_tPtr],
                                            function (gdp_event) {
                                            if (typeof gdp_event !== 'undefined') {
                                                // Here, we want to first get the datum.
                                                var datum = gdpjs.gdp_event_getdatum_js(gdp_event);
                                                var data = gdpjs.gdp_datum_buf_as_string(datum);
                                                // console.log("gdp.js: callback: data is: ");
                                                self.emit("data", data);
                                                
                                                // gdp/gdp/doc/gdp-programmatic-api.html
                                                // says that for gdp_gcl_subscribe(), "It is
                                                // the responsibility of the callback function
                                                // to call gdp_event_free(gev)."
                                                gdpjs.gdp_event_free_js(gdp_event);
                                            }
                                        });

    console.log("\nFIXME: accessors/web/hosts/node/node_modules/gdp/gdp.js: subscribe(): Need to eventually remove subscription callbacks from the list of references.");
    _subscriptionCallBacks.push(gdp_event_cbfunc);
    
    var data = gdpjs.read_gcl_records(gdpd_addr, gcl_name,
                                      gcl_firstrec, gcl_numrecs,
                                      gcl_subscribe, gcl_multiread, /*recdest,*/
                                      conout, gdp_event_cbfunc,
                                      wait_for_events,
                                      gcl_get_next_event
                                     );
    //console.log("node_modules/gdp/gdp.js:subscribe(): returned from read_gcl_records(). data:");
    //console.log(data);
    //if (data.records.length > 0) {
    //    return data.records[0].value;
    //} else {
    //    return null;
    //}

};

exports.GDP.prototype.unsubscribe = function () {
    this.close();
    console.log("FIXME: accessors/web/hosts/node/node_modules/gdp/gdp.js: Need unsubscribe()");
}
